ostd/src/arch/riscv/timer/mod.rs

```diff
@@ -2,48 +2,116 @@
-/// a modest choice.
+/// The timer frequency (Hz). Here we choose 1000Hz since 1000Hz is easier for
+/// unit conversion and convenient for timer. What's more, the frequency cannot
+/// be set too high or too low, 1000Hz is a modest choice.
 ///
-/// For system performance reasons, this rate cannot be set too high, otherwise most of the time
-/// is spent executing timer code.
+/// For system performance reasons, this rate cannot be set too high, otherwise
+/// most of the time is spent executing timer code.
 pub const TIMER_FREQ: u64 = 1000;
 
-pub(crate) static TIMEBASE_FREQ: AtomicU64 = AtomicU64::new(1);
-
-/// [`IoMem`] of goldfish RTC, which will be used by `aster-time`.
-pub static GOLDFISH_IO_MEM: Once<IoMem> = Once::new();
+static TIMEBASE_FREQ: AtomicU64 = AtomicU64::new(0);
+static TIMER_INTERVAL: AtomicU64 = AtomicU64::new(0);
 
 pub(super) fn init() {
-    let timer_freq = DEVICE_TREE
-        .get()
-        .unwrap()
-        .cpus()
-        .next()
-        .unwrap()
-        .timebase_frequency() as u64;
-    TIMEBASE_FREQ.store(timer_freq, Ordering::Relaxed);
-
-    let chosen = DEVICE_TREE.get().unwrap().find_node("/soc/rtc").unwrap();
-    if let Some(compatible) = chosen.compatible()
-        && compatible.all().any(|c| c == "google,goldfish-rtc")
-    {
-        let region = chosen.reg().unwrap().next().unwrap();
-        let io_mem = unsafe {
-            IoMem::new(
-                (region.starting_address as usize)
-                    ..(region.starting_address as usize) + region.size.unwrap(),
-                crate::mm::page_prop::PageFlags::RW,
-                crate::mm::page_prop::CachePolicy::Uncacheable,
-            )
-        };
-        GOLDFISH_IO_MEM.call_once(|| io_mem);
+    TIMEBASE_FREQ.store(
+        DEVICE_TREE
+            .get()
+            .unwrap()
+            .cpus()
+            .next()
+            .unwrap()
+            .timebase_frequency() as u64,
+        Ordering::Relaxed,
+    );
+    TIMER_INTERVAL.store(
+        TIMEBASE_FREQ.load(Ordering::Relaxed) / TIMER_FREQ,
+        Ordering::Relaxed,
+    );
+
+    if is_sstc_enabled() {
+        // SAFETY: Mutating the static variable `SET_NEXT_TIMER_FN` is safe here
+        // because we ensure that it is only modified during the initialization
+        // phase of the timer.
+        unsafe {
+            SET_NEXT_TIMER_FN = set_next_timer_sstc;
+        }
+    }
+    set_next_timer();
+    // SAFETY: Accessing the `sie` CSR to enable the timer interrupt is safe
+    // here because this function is only called during timer initialization,
+    // and we ensure that only the timer interrupt bit is set without affecting
+    // other interrupt sources.
+    unsafe {
+        riscv::register::sie::set_stimer();
+    }
+}
+
+pub(super) fn handle_timer_interrupt() {
+    let irq_guard = trap::disable_local();
+    if irq_guard.current_cpu() == CpuId::bsp() {
+        crate::timer::jiffies::ELAPSED.fetch_add(1, Ordering::SeqCst);
```

Comment:
I can't think of any race conditions that would be caused if `SeqCst` is weaned to `Relaxed`, can you?